examples/factory/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+    ///   contract instance. Otherwise [`FactoryError::NonExistentProduct`] is
+    ///   returned.
+    ///
+    /// * The supplied instance *must* be an instance of the `product` contract
+    ///   defined in this module. Otherwise [`FactoryError::InvalidProduct`] is
+    ///   returned.
+    ///
+    /// * The supplied instance *must* be uninitialized. Moreover, the
+    ///   preconditions for [`crate::product::initialize`] must be met. (In
+    ///   particular, the invoker of this transaction must be the same account
+    ///   as created the product instance.) Otherwise
+    ///   [`FactoryError::InitializeFailed`] is returned.
+    ///
+    /// If successful, the product is assigned a fresh index (as part of the
+    /// initialization), and is recorded in the factory's index.
+    #[receive(contract = "factory", name = "produce", parameter = "ContractAddress", mutable)]
+    pub fn produce(
+        ctx: &ReceiveContext,
+        host: &mut Host<FactoryState>,
+    ) -> Result<(), FactoryError> {
+        let product_address = ctx.parameter_cursor().get()?;
+        // We can depend upon getting the module reference for our own contract.
+        let self_module_ref = host.contract_module_reference(ctx.self_address()).unwrap();
+        // Check the product module is the same as our own module.
+        let product_module_ref = host
+            .contract_module_reference(product_address)
+            .or(Err(FactoryError::NonExistentProduct))?;
+        if self_module_ref != product_module_ref {
+            Err(FactoryError::InvalidProduct)?;
+        }
+        // Check the product contract name is correct.
+        // As this module contains both the factory and product contracts, it is not
+        // enough to check just the module reference, because we want to ensure
+        // that the would-be product is not an instance of the factory contract.
+        let product_name =
+            host.contract_name(product_address).or(Err(FactoryError::NonExistentProduct))?;
+        if product_name != PRODUCT_INIT_NAME {
+            Err(FactoryError::InvalidProduct)?;
+        }
+        // Invoke the initialize entrypoint on the product passing in the index for this
+        // product.
+        let next_product = host.state().next_product;
+        host.invoke_contract(
+            &product_address,
+            &next_product,
+            EntrypointName::new_unchecked(PRODUCT_INITIALIZE_ENTRYPOINT),
+            Amount::zero(),
+        )
+        .or(Err(FactoryError::InitializeFailed))?;
+        // Update the state
+        let state = host.state_mut();
+        state.next_product = next_product + 1;
+        state.products.insert(next_product, product_address);
+        Ok(())
+    }
+
+    /// Get the number of product instances that have been successfully
+    /// initialized by this factory.
+    #[receive(contract = "factory", name = "view_product_count")]
```

Comment:
Can you add `return_value`
```suggestion
    #[receive(contract = "factory", name = "view_product_count", return_value = "u64")]
```

Not feeling strongly about it: With regards to the errors, we can keep it as it is to keep the contract simple or we can add.
```suggestion
    #[receive(contract = "factory", name = "view_product_count", error = "FactoryError")]
```